rust/lance-encoding/src/encodings/physical/binary.rs

```diff
@@ -0,0 +1,421 @@
+    ) {
+        println!("Inside BinaryPageDecoder update_capacity");
+        println!("Rows to skip: {:?}, Num rows: {:?}", rows_to_skip, num_rows);
+        let offsets = self
+            .decoded_indices
+            .as_any()
+            .downcast_ref::<UInt32Array>()
+            .unwrap(); // 4, 9, 13, 18, 23, 27
+        let (bytes_to_skip, num_bytes, _) = get_bytes_from_rows(rows_to_skip, num_rows, offsets);
+
+        // 32 bits or 4 bytes per value.
+        buffers[0].0 = (num_rows as u64) * 4;
+        buffers[0].1 = true;
+
+        println!(
+            "Capacity update: Bytes to skip: {:?}, Num bytes: {:?}",
+            bytes_to_skip, num_bytes
+        );
+        self.bytes_decoder
+            .update_capacity(bytes_to_skip, num_bytes, &mut buffers[1..], all_null);
+    }
+
+    fn decode_into(
+        &self,
+        rows_to_skip: u32,
+        num_rows: u32,
+        dest_buffers: &mut [bytes::BytesMut],
+    ) -> Result<()> {
+        println!("Rows to skip: {:?}, Num rows: {:?}", rows_to_skip, num_rows);
+        let offsets = self
+            .decoded_indices
+            .as_any()
+            .downcast_ref::<UInt32Array>()
+            .unwrap(); // 4, 9, 13, 18, 23, 27
+        println!("Offsets: {:?}", offsets);
+
+        let (bytes_to_skip, num_bytes, target_offsets) =
+            get_bytes_from_rows(rows_to_skip, num_rows, offsets);
+
+        println!("Target offsets: {:?}", target_offsets);
+        println!(
+            "Bytes to skip: {:?}, Num bytes: {:?}",
+            bytes_to_skip, num_bytes
+        );
+
+        // copy target_offsets into dest_buffers[0]
+        let values = target_offsets.values();
+        let byte_slice: &[u8] = unsafe {
+            std::slice::from_raw_parts(
+                values.as_ptr() as *const u8,
+                values.len() * std::mem::size_of::<u32>(),
+            )
+        };
+
+        dest_buffers[0].extend_from_slice(byte_slice);
+        self.bytes_decoder
+            .decode_into(bytes_to_skip, num_bytes, &mut dest_buffers[1..])?;
+
+        // for i in 0..3 {
+        //     println!(
+        //         "length, capacity of buffer: {:?}, {:?} ",
+        //         dest_buffers[i].len(),
+        //         dest_buffers[i].capacity()
+        //     );
+        // }
+
+        Ok(())
+    }
+
+    fn num_buffers(&self) -> u32 {
+        self.bytes_decoder.num_buffers() + 1
+        // 0
+    }
+}
+
+#[derive(Debug)]
+pub struct BinaryEncoder {
+    indices_encoder: Box<dyn ArrayEncoder>,
+    bytes_encoder: Box<dyn ArrayEncoder>,
+}
+
+impl BinaryEncoder {
+    pub fn new(
+        indices_encoder: Box<dyn ArrayEncoder>,
+        bytes_encoder: Box<dyn ArrayEncoder>,
+    ) -> Self {
+        Self {
+            indices_encoder,
+            bytes_encoder,
+        }
+    }
+}
+
+// Returns indices arrays from string arrays
+// These arrays are padded with a 0 at the beginning
+fn get_indices_from_string_arrays(arrays: &[ArrayRef]) -> Vec<ArrayRef> {
```

Comment:
This logic isn't quite right when you have multiple arrays.  For example, let's assume you have:

```
[
  ["foo", "bar"],
  ["baz"],
]
```

In this case you will have incoming offsets of:

```
[
  [0, 3, 6],
  [0, 3]
]
```

The logic here will output `[3, 6, 3]`.  What you want is either `[0, 3, 6, 9]` or `[3, 6, 9]`.